### ISSUE TITLE
Copy lists behavior when including folders not well defined

### ISSUE THREAD

**Issue by ramirezfranciscof** [aiidateam/aiida-core#4499]
After the last re-structure of the way the copy lists work (#4416), I'm having some problems running PpCalcJobs in [aiida-qe](https://github.com/aiidateam/aiida-quantumespresso). I think this slightly changed the behavior of including whole folders in the copy lists that affected how PP copied files from the `parent_folder` but not PW. I printed the source and target dirs and noticed this (notice the asterisc):

```
This works (Pw):
aiida/2e/bb/e947-1fda-48a5-9085-e1fff307cc28/./out/*  =>  ./out/

This breaks (Pp):
aiida/30/26/791d-5880-40c7-b0df-4fd78b291dcb/./out/  =>  ./out/
```

I made the following pytest if someone wants to try it (there are many commented examples I started thinking of in `general_copy_list` that I think should be discussed/considered, but the two separated ones are the relevant for this particular break): 

```python
import io
import os
import pathlib
import pytest

from aiida import orm
from aiida.engine.daemon import execmanager
from aiida.common.datastructures import CalcInfo, CodeInfo
from aiida.transports.plugins.local import LocalTransport

def test_upload_calculation(aiida_localhost, aiida_local_code_factory, fixture_sandbox, tmp_path):
    """Test folders in the copy lists."""

    # FolderData: needs to be stored when is checked by upload_calculation
    folder_node = orm.FolderData()
    folder_node.put_object_from_filelike(io.StringIO('dummy_content'), 'subfolder/file.txt')
    folder_node.store()

    # RemoteData: upload_calculation will not check if it is stored (should it?)
    remote_node = orm.RemoteData(computer=aiida_localhost, remote_path=str(tmp_path))

    remote_path_dir = tmp_path / 'subfolder'
    remote_path_dir.mkdir()
    remote_path = str(tmp_path / 'subfolder' / 'file.txt')
    with open(remote_path, 'w') as handle:
        handle.write('dummy_content')
        handle.flush()

    # All of these should work correctly?...
    general_copy_list = []
    #general_copy_list.append( ('subfolder', 'subfolder') )
    #general_copy_list.append( ('subfolder/', 'subfolder') )
    #general_copy_list.append( ('subfolder/*', 'subfolder') )
    #general_copy_list.append( ('./subfolder', 'subfolder') )

    # UNCOMMENT THIS
    #general_copy_list.append( ('./subfolder/', './subfolder/') )
    general_copy_list.append( ('./subfolder/*', './subfolder/') )
    copy_list_range = ["local", "remote", "symlink"]
    copy_list_range = ["remote"]

    folder_uuid = folder_node.uuid
    remote_uuid = remote_node.computer.uuid
    for copy_src0, copy_tgt in general_copy_list:
      for copy_list in copy_list_range:

        # CodeInfo and CalcInfo: need to be set up as normal for any CalcJob, but we also
        # need to manually set calc_info.uuid  as we are skipping the step of the engine
        # were this happens.
        code_node = aiida_local_code_factory('arithmetic.add', '/bin/bash')
        code_info = CodeInfo()
        code_info.code_uuid = code_node.uuid
        calc_node = orm.CalcJobNode(computer=aiida_localhost).store()
        calc_info = CalcInfo()
        calc_info.uuid = calc_node.uuid
        calc_info.codes_info = [code_info]

        copy_full = os.path.join(remote_node.get_remote_path(), copy_src0)
        if copy_list == "local":
            calc_info.local_copy_list     = [(folder_uuid, copy_src0, copy_tgt)]
        if copy_list == "remote":
            calc_info.remote_copy_list    = [(remote_uuid, copy_full, copy_tgt)]
        if copy_list == "symlink":
            calc_info.remote_symlink_list = [(remote_uuid, copy_full, copy_tgt)]

        # We need to manually open a transport and pass it to upload_calculation, together
        # with a pre-set up sandbox folder (in this case it can be empty since we are mostly
        # checking through the copy lists)
        with LocalTransport() as transport:
            execmanager.upload_calculation(calc_node, transport, calc_info, fixture_sandbox)

        # Check the folder and file is there
        calc_folder_path = pathlib.Path(calc_node.get_remote_workdir())

        folder_list = calc_folder_path.iterdir()
        folder_path = calc_folder_path / 'subfolder'
        assert list(folder_list) == [folder_path]

        file_list = (calc_folder_path / 'subfolder').iterdir()
        file_path = calc_folder_path / 'subfolder' / 'file.txt'
        assert list((calc_folder_path / 'subfolder').iterdir()) == [calc_folder_path / 'subfolder/file.txt']
        assert list(file_list) == [file_path]

        full_file_path = str(calc_folder_path / 'subfolder/file.txt')
        with open(full_file_path, 'r') as handle:
            assert handle.read() == 'dummy_content'
```

**Comment by ramirezfranciscof** [comment]
@sphuber I'm not sure why the nightly build of aiida-qe is not catching this, maybe I have and old version and in the more up-to date PP copies files the same way as PW? In any case the pytest with aiidacore raises so it could potentially break other plugins.

**Comment by sphuber** [comment]
> 'm not sure why the nightly tests of aiida-qe are not catching this,

Well, if the bug is really in the `upload_calculation`, then it won't be caught, because that method is actually never called by the test suite of `aiida-quantumespresso`. The reason is that we don't *actually* want to run the calculation job, because we don't have an actual installation of Quantum ESPRESSO, plus we are not interested in it actually running. So far, the tests are simply running the `prepare_for_submission` because that is also the code that belongs to `aiida-quantumespresso` and so is the part that is its responsibility to test. Still, `upload_calculation` contains a lot of logic and has been causing problems before, so maybe we should try and find a way to test this as well in `aiida-quantumespresso`, even though this really should just be the responsibility of `aiida-core`.

**Comment by greschd** [comment]
Just to mention, if you want to add "real" calculations to the `aiida-quantumespresso` test suite, I think the proper way to go would be using `aiida-testing`: https://github.com/aiidateam/aiida-testing

There are [some issues we need to address](https://github.com/aiidateam/aiida-testing/issues?q=is%3Aopen+is%3Aissue+milestone%3A%22Version+0.1+release%22) before releasing a `0.1` there; and unfortunately I've neglected it a bit because I'm not actively working on anything that'd need it. But maybe together we can push it over the line @sphuber @ltalirz.

**Comment by sphuber** [comment]
Since you don't specify exactly what kind of `PpCalculation` you are running and how it fails, i.e. show the exception and stack trace, there is not much I can say directly here.

That being said, looking at the current code of the transports and the copy lists, there are quite a number of problems that have to do with:
1) Not all transport plugins always doing the same thing
2) The interface of the copy lists not being similar and therefore difficult to understand and use
3) Lack of testing on `aiida-core` side

To the first point, the reason you may see things fail on your side but not on `aiida-quantumespresso` is maybe you are using a `LocalTransport` whereas most uses of `aiida-quantumespresso` use `SshTransport`. The example you give of the instruction `('./out/', './out/')` fails for the `LocalTransport` but it succeeds for the `SshTransport`. Now when I say succeed, the actual transport copy operation works, but it does the incorrect thing with our current implementation of `upload_calculation`, which goes back to point three that we need more testing.

To the second point, the idea of the copy lists is non-ideal I think. For the remote lists there might not be a better solution, however, for the `local_copy_list` I really think we should deprecate it and have people simply use the sandbox folder in combination with the `provenance_exclude_list`. As to its unintuitiveness, from your example, it seems you think the local list has the same syntax as the remote ones, which is not true. Where the source path can contain wildcards for the remote lists, the local list *has* to have a fully defined (relative) path.

**Comment by ramirezfranciscof** [comment]
I think the whole copy list rethink is a worthy topic but perhaps not the one I indended to discuss here. I wanted to focus more on solving the concrete problem of not breaking the current interface now, maybe later think how to improve said interface. Maybe the general design of my test is not the best and led to this detour, but notice that (although, as you pointed out, I tried incorrectly to design it more generally to test all lists) for now it is only testing the `remote_copy_list`.

The approach I thought for this is to be explicit here about how the different copying syntax for source and target. If we compile a list I can check how they behaved previous to the change and see how that changed now and what can be done to recover the previous behavior. Moreover, we could decide later on a *correct* behavior and implement it (issuing warnings of deprecation for a couple of versions before doing so). For example, I would try to have all transports behave in the same way (which I'm not sure was the case prior to the update).

**Comment by sphuber** [comment]
I agree that it is a good idea to see what the previous behavior was for the various copy lists, especially with respect to the behavior of overlapping paths, where some of the subfolders may already exist when executing a copy. Maybe would be good to go back to v1.2 because I think we have touched the functionality in `upload_calculation` a few times, not just in v1.4. Maybe go a bit through the history of the method to see what other versions may have affected its behavior.

**Comment by ramirezfranciscof** [comment]
Closed in favour of #4566
